driver/bk4819.c

```diff
@@ -1281,54 +1273,54 @@ void BK4819_EnableTXLink(void)
 
 void BK4819_PlayDTMF(char Code)
 {
-	uint16_t tone1 = 0;
-	uint16_t tone2 = 0;
-
-	switch (Code)
-	{
-		case '0': tone1 = 941; tone2 = 1336; break;
-		case '1': tone1 = 679; tone2 = 1209; break;
-		case '2': tone1 = 697; tone2 = 1336; break;
-		case '3': tone1 = 679; tone2 = 1477; break;
-		case '4': tone1 = 770; tone2 = 1209; break;
-		case '5': tone1 = 770; tone2 = 1336; break;
-		case '6': tone1 = 770; tone2 = 1477; break;
-		case '7': tone1 = 852; tone2 = 1209; break;
-		case '8': tone1 = 852; tone2 = 1336; break;
-		case '9': tone1 = 852; tone2 = 1477; break;
-		case 'A': tone1 = 679; tone2 = 1633; break;
-		case 'B': tone1 = 770; tone2 = 1633; break;
-		case 'C': tone1 = 852; tone2 = 1633; break;
-		case 'D': tone1 = 941; tone2 = 1633; break;
-		case '*': tone1 = 941; tone2 = 1209; break;
-		case '#': tone1 = 941; tone2 = 1477; break;
-	}
-
-	if (tone1 > 0)
-		BK4819_WriteRegister(BK4819_REG_71, (((uint32_t)tone1 * 103244) + 5000) / 10000);   // with rounding
-	if (tone2 > 0)
-		BK4819_WriteRegister(BK4819_REG_72, (((uint32_t)tone2 * 103244) + 5000) / 10000);   // with rounding
+	int32_t index = ((Code >= 65) ? (Code - 55) : ((Code <= 35) ? 15 :((Code <= 42) ? 14 : (Code - '0'))));
```

Comment:
This index doesn't respect the previous checks. You're converting all invalid values below star to be either star or hash, instead of not playing them at all.

The index shouldn't need to be signed either, which leads to your check >= 0.

Overall the entire line is much less readable, even if you changed the numeri so values by 'A', etc.

Reply:
The index couldn't overflow the array size (there's a check before writing the register value).
The idea is to remap the **signed** char to an index, and keep it to the boundaries, and also avoiding a huge switch/case, which is a long "if (x) {} else if (y) {}....."
Of course, if the function prototype wasn't passing a char, but some king of DTMF enum, it would be even better.
The only way I see a problem is if the "codeplug" get messed.